[VectorDistribution] Add pattern to distribute contractions

harsh-nod opened the pull request:
This PR adds a pattern to distribute contractions to MFMA ops. Currently, 16x16x16 and 32x32x8 MFMA operations are supported.

harsh-nod committed:
[VectorDistribution] Add pattern to distribute contractions

This PR adds a pattern to distribute contractions to
MFMA ops. Currently, 16x16x16 and 32x32x8 MFMA operations
are supported.

Groverkss commented on compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:
```diff
@@ -929,32 +929,6 @@ void transform_dialect::WorkgroupSwizzleOp::getEffects(
 // TestVectorLayoutAnalysisOp
 //===----------------------------------------------------------------------===//
 
-static void setAnchorOpsFromAttributes(VectorLayoutAnalysis &analysis,
```
I prefer this staying in TransformExtensions tests. Is there any reason we shifted this?

harsh-nod replied:
Yes, I needed a way to set layouts from annotations in the IR and this function was "trapped" inside the  cpp code. What I was imagining was that the user (either through a separate pass or some other way) would set layouts on specific ops and then we would read the layouts and propagate to the rest of the operations. This function here is the first step towards that goal. I didn't want to expand the scope of the PR so I settled for just moving it to a place where it can be used by others.

Groverkss replied:
SGTM

Groverkss commented on compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:
```diff
@@ -1496,5 +1500,41 @@ void transform_dialect::PackSharedMemoryAllocOp::getEffects(
+  int64_t k = getK();
+  MFMAType mfmaType;
+  if ((m == 16) && (n == 16) && (k == 16)) {
+    mfmaType = MFMAType::F16_16x16x16_F32;
```
Is it not possible to determine the MFMA type from the layout? This looks like something the layout should tell us.

harsh-nod replied:
For the 3 different matrix cores we have seen so far, I would agree. But in general, you can imagine a scenario where different instructions use the same layout. It seems like having the ability to infer a set of compatible instructions for a given layout does makes sense though, acting kind of like a verifier. I can modify this to reflect that.

Groverkss commented on compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:
```diff
@@ -1496,5 +1500,41 @@ void transform_dialect::PackSharedMemoryAllocOp::getEffects(
+  } else {
+    mfmaType = MFMAType::F16_32x32x8_F32;
+  }
+  populateAMDGPUDistributionPatterns(patterns, mfmaType);
```
Is it not possible for the patterns to match on the layout to check what mfmaType to use? I don't like passing mfmaType to patterns.

harsh-nod replied:
Yes, I think this is possible.

Groverkss commented on compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensionsOps.td:
```diff
@@ -682,4 +682,44 @@ def PackSharedMemoryAllocOp :  Op<Transform_Dialect, "iree.pack_shared_memory_al
+    let arguments = (ins TransformHandleTypeInterface:$target,
+                         I64Attr:$m,
+                         I64Attr:$n,
+                         I64Attr:$k);
```
I feel like the layout should contain enough information to find m, n, k.

harsh-nod replied:
For our use case, yes, but in general not necessarily.

Groverkss commented on compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp:
```diff
@@ -0,0 +1,168 @@
+  DistributeContractions(MLIRContext *ctx, MFMAType mfmaType)
+      : OpDistributionPattern(ctx), mfmaType(mfmaType) {}
+
+  SmallVector<int64_t> getIndices(ContractType contractType,
```
Can we add some documentation on what this does?

harsh-nod replied:
Sure I will add some comments.

Groverkss commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/VectorExt/IR/VectorExtOps.h:
```diff
@@ -70,27 +69,70 @@ class DimensionalRange {
-      llvm::DenseMap<int64_t, SmallVector<LayoutDimension>>;
-  void maybeFreezeAndConcatenate(const LayoutIterator &frozenIterator);
+  struct State {
+    SmallVector<int64_t> computeSIMTIndex() const;
```
I think we have this implemented in GPUDistributionPatterns.cpp also

Groverkss replied:
Maybe we can remove that?

harsh-nod replied:
yes looks to be a duplicate. I will remove that.

Groverkss commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/VectorExt/IR/VectorExtOps.h:
```diff
@@ -70,27 +69,70 @@ class DimensionalRange {
 // required during distribution.
 class LayoutIterator {
```
Can we move this out of VectorExtOps.h into it's own file?

harsh-nod replied:
Sure, but what's the reason for doing this? There's nothing in VectorExtOps.h except for the LayoutIterator class and classes that it uses.

Groverkss replied:
It isn't really related to VectorExtOps at all. But let's do it in a seperate patch.

harsh-nod replied:
Will do thanks

Groverkss commented on llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtDialect.cpp:
```diff
@@ -109,17 +109,41 @@ VectorLayoutInterface LayoutAttr::permute(ArrayRef<int64_t> permutation) const {
-  LayoutDimension labels[] = {LayoutDimension::BATCHX, LayoutDimension::BATCHY,
-                              LayoutDimension::VECTORX};
+SmallVector<int64_t> LayoutAttr::getDistributedShape() const {
+  SmallVector<LayoutDimension> labels{
```
Can you add a comment here that this is the order in which the shape is distributed?

harsh-nod replied:
Sure, will add.

Groverkss reviewed (changes requested):
Thank you for this patch. I added some initial comments.

harsh-nod committed:
Distribute to MFMA only on valid layouts

Adds the ability to infer the MFMA type.
Distribution is done only if the operands have
valid MFMA layouts. Also adds batch test.

kuhar commented on compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp:
```diff
@@ -1019,3 +1019,33 @@ void VectorLayoutAnalysis::dump() {
+    for (NamedAttribute attr : op->getAttrs()) {
+      StringRef name = attr.getName().strref();
+      if (name.find("__vector_layout_test_anchor_operand_") !=
+          std::string::npos) {
```
```suggestion
      if (name.contains("__vector_layout_test_anchor_operand_")) {
```

harsh-nod replied:
thanks! will change.

kuhar commented on compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp:
```diff
@@ -1019,3 +1019,33 @@ void VectorLayoutAnalysis::dump() {
+        analysis.setAnchor(op->getOperand(operandNum), attr.getValue());
+      }
+      if (name.find("__vector_layout_test_anchor_result_") !=
+          std::string::npos) {
```
```suggestion
      if (name.contains("__vector_layout_test_anchor_result_")) {
```

harsh-nod replied:
This one too.

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp:
```diff
@@ -0,0 +1,355 @@
+      return ContractType::MTM;
+    } else if ((maps == infer({{k, m}, {n, k}, {m, n}})) ||
+               (maps == infer({{k, n}, {m, k}, {n, m}}))) {
+      return ContractType::MTMT;
```
nit: no else after return per https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

harsh-nod replied:
thanks. that reads better too.

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp:
```diff
@@ -0,0 +1,355 @@
+    auto colLayout = createPerDimLayout(
+        ctx, {batchColLabel, LayoutDimension::LANEY, LayoutDimension::VECTORX},
+        {batchCol, 4, 4});
+    return std::make_tuple(rowLayout, colLayout);
```
nit:
```suggestion
    return {rowLayout, colLayout};
```

harsh-nod replied:
thanks. will change.

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp:
```diff
@@ -0,0 +1,355 @@
+    MLIRContext *ctx = getContext();
+    auto rowLayout = createPerDimLayout(
+        ctx, {batchRowLabel, LayoutDimension::LANEX}, {batchRow, 16});
+    auto colLayout = createPerDimLayout(
```
nit: could you write the full type when it's not obvious based on the RHS only?

harsh-nod replied:
sure I can add the types.

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp:
```diff
@@ -0,0 +1,355 @@
+                                LayoutDimension batchColLabel, int64_t batchCol,
+                                ContractMatrixType matrixType) const {
+    MLIRContext *ctx = getContext();
+    auto rowLayout = createPerDimLayout(
```
nit: also here

harsh-nod replied:
sure will change.

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp:
```diff
@@ -0,0 +1,355 @@
+          {batchColLabel, LayoutDimension::LANEY, LayoutDimension::VECTORX},
+          {batchCol, 2, 4});
+    }
+    return std::make_tuple(rowLayout, colLayout);
```
```suggestion
    return {rowLayout, colLayout};
```

harsh-nod replied:
will do

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp:
```diff
@@ -0,0 +1,355 @@
+    SmallVector<VectorValue> operands;
+    SmallVector<LayoutAttr> layouts;
+    for (auto [operand, layout] :
+         llvm::zip(contractOp->getOperands(), signature.operands)) {
```
Use zip_equals when you expect both ranges to have equal lengths

harsh-nod replied:
sure will change to use that.

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/test/amdgpu_contraction_distribution.mlir:
```diff
@@ -0,0 +1,115 @@
+    transform.iree.test_amdgpu_contraction_distribution %top_level_func : !transform.any_op
+    transform.yield
+  }
+}
```
nit: missing newline

harsh-nod replied:
will add

kuhar commented on llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtDialect.cpp:
```diff
@@ -109,17 +109,46 @@ VectorLayoutInterface LayoutAttr::permute(ArrayRef<int64_t> permutation) const {
+std::optional<int64_t> LayoutAttr::getBatchDim(int64_t dim) {
+  assert(dim < getLayouts().size());
+  PerDimLayoutAttr layout = getDimLayout(dim);
+  for (auto [name, shape] : llvm::zip(layout.getLabels(), layout.getShapes())) {
```
also here: zip_equal if we can

harsh-nod replied:
yes can change here as well.

kuhar commented on llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtOps.cpp:
```diff
@@ -52,81 +52,158 @@ OpFoldResult ToSIMTOp::fold(FoldAdaptor) {
+  DenseMap<LayoutDimension, int64_t> strides;
+  for (auto perDimAttr : llvm::enumerate(attr.getLayouts())) {
+    initialize(perDimAttr.value(), strides, perDimAttr.index());
+  }
```
```suggestion
  for (auto [idx, attr] : llvm::enumerate(attr.getLayouts())) {
    initialize(attr, strides, idx);
  }
```

harsh-nod replied:
nice! this is much cleaner.

kuhar commented on llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtOps.cpp:
```diff
@@ -52,81 +52,158 @@ OpFoldResult ToSIMTOp::fold(FoldAdaptor) {
+    if (perDimAttr.index() != simtIndex)
+      continue;
+    initialize(perDimAttr.value(), strides, perDimAttr.index());
+  }
```
```suggestion
  for (auto [idx, attr] : llvm::enumerate(attr.getLayouts())) {
    if (idx != simtIndex)
      continue;
    initialize(attr, strides, idx);
  }
```

harsh-nod replied:
will change.

kuhar commented on llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtOps.cpp:
```diff
@@ -52,81 +52,158 @@ OpFoldResult ToSIMTOp::fold(FoldAdaptor) {
+    if (perDimAttr.index() != simtIndex)
+      continue;
+    initialize(perDimAttr.value(), strides, perDimAttr.index());
   }
```
```suggestion
  for (auto [idx, attr] : llvm::enumerate(attr.getLayouts())) {
    if (idx != simtIndex)
      continue;
    initialize(attr, strides, idx);
  }
```

harsh-nod replied:
same.

kuhar commented on llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtOps.cpp:
```diff
@@ -52,81 +52,158 @@ OpFoldResult ToSIMTOp::fold(FoldAdaptor) {
+    break;
   }
-  return complete;
+  iterations++;
```
```suggestion
  ++iterations;
```

harsh-nod replied:
thanks forgot about this one https://llvm.org/docs/CodingStandards.html#id57

kuhar commented on llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtOps.cpp:
```diff
@@ -52,81 +52,158 @@ OpFoldResult ToSIMTOp::fold(FoldAdaptor) {
+LayoutIterator::State::computeIteratorProjectedSIMTIndex() const {
+  SmallVector<int64_t> indices = computeSIMTIndex();
+  SmallVector<int64_t> projectedIndices;
+  for (int i = 0; i < labels.size(); i++) {
```
```suggestion
  for (size_t i = 0, e = labels.size(); i != e; ++i) {
```

harsh-nod replied:
thanks will modify.

kuhar reviewed (commented):
Just a few drive-by nits

harsh-nod committed:
Conform to LLVM formatting guidelines

Groverkss commented on compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_vector_distribution.mlir:
```diff
@@ -7,31 +7,31 @@ builtin.module attributes { transform.with_named_sequence } {
   func.func @distribute_elementwise_f16(%a: vector<16x16xf16>, %b: vector<16x16xf16>) -> vector<16x16xf16> {
     %c0 = arith.constant 0 : index
     %cst_0 = arith.constant 0.0 : f16
-    // CHECK: %[[ROOT:.*]] = arith.constant dense<0.000000e+00> : vector<4xf16>
```
Shouldn't the distributed type be 4x4xf16? Why 16xf16?

harsh-nod replied:
Good question. So at least for MFMA, all the vector dimensions need to be grouped together in the SIMT distributed vector. So here you have vectorx and vectory which is why it is 16xf16.

Groverkss commented on compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensionsOps.td:
```diff
@@ -682,4 +682,41 @@ def PackSharedMemoryAllocOp :  Op<Transform_Dialect, "iree.pack_shared_memory_al
   }];
 }
 
+def TestAMDGPUContractionDistribution :
```
Flyby comment, we should later just make this TestAMDGPUDistributionPatterns or something. Nothing to do for now.

harsh-nod replied:
sounds good

Groverkss reviewed (commented):
The only concern I have are the tests. I think something is wrong because while it is a fix, the distributed shape should be 4x4

Groverkss reviewed (approved):
LGTM
